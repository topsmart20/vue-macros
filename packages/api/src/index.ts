export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
