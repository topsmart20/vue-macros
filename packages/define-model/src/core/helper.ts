export const helperPrefix = '/plugin-define-model'

export const emitHelperId = `${helperPrefix}/emit-helper`
export const emitHelperCode = `export default (emitFn, key, value, ...args) => {
  emitFn(key, value)
  return args.length > 0 ? args[0] : value
}`

export const useVmodelHelperId = `${helperPrefix}/use-vmodel`
export const useVmodelHelperCode = `import { getCurrentInstance } from 'vue';
import { useVModel } from '@vueuse/core';
export default (...keys) => {
  const props = getCurrentInstance().proxy.$props
  const ret = {}
  for (const [key, eventName] of keys)
    ret[key] = useVModel(props, key, undefined, { eventName })
  return ret
}
`
