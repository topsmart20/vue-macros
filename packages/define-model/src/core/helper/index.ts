export const helperPrefix = '/plugin-define-model'

export const emitHelperId = `${helperPrefix}/emit-helper`
export { default as emitHelperCode } from './emit-helper?raw'

export const useVmodelHelperId = `${helperPrefix}/use-vmodel`
export { default as useVmodelHelperCode } from './use-vmodel?raw'
