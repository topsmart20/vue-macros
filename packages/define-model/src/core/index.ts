import { extractIdentifiers, walkAST } from 'ast-walker-scope'
import {
  DEFINE_EMITS,
  DEFINE_MODEL,
  DEFINE_OPTIONS,
  DEFINE_PROPS,
  MagicString,
  REPO_ISSUE_URL,
  getTransformResult,
  isCallOf,
  parseSFC,
} from '@vue-macros/common'
import { emitHelperId } from './helper'
import type {
  Identifier,
  LVal,
  Node,
  ObjectExpression,
  ObjectPattern,
  ObjectProperty,
  Program,
  Statement,
  TSInterfaceBody,
  TSTypeLiteral,
  VariableDeclaration,
} from '@babel/types'

export const transformDefineModel = (
  code: string,
  id: string,
  version: 2 | 3
) => {
  let hasDefineProps = false
  let hasDefineEmits = false
  let hasDefineModel = false
  let propsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let propsDestructureDecl: Node | undefined
  let emitsTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let emitsIdentifier: string | undefined
  let modelTypeDecl: TSInterfaceBody | TSTypeLiteral | undefined
  let modelIdentifier: string | undefined
  let modelDeclKind: string | undefined
  let modelDestructureDecl: ObjectPattern | undefined
  const modelIdentifiers = new Set<Identifier>()
  const modelVue2: { event: string; prop: string } = { prop: '', event: '' }

  function processDefinePropsOrEmits(node: Node, declId?: LVal) {
    let type: 'props' | 'emits'
    if (isCallOf(node, DEFINE_PROPS)) {
      type = 'props'
    } else if (isCallOf(node, DEFINE_EMITS)) {
      type = 'emits'
    } else {
      return false
    }
    const fnName = type === 'props' ? DEFINE_PROPS : DEFINE_EMITS

    if (type === 'props') hasDefineProps = true
    else hasDefineEmits = true

    if (node.arguments[0])
      throw new SyntaxError(
        `${fnName}() cannot accept non-type arguments when used with ${DEFINE_MODEL}()`
      )

    const typeDeclRaw = node.typeParameters?.params?.[0]
    if (!typeDeclRaw)
      throw new SyntaxError(
        `${fnName}() expected a type parameter when used with ${DEFINE_MODEL}.`
      )

    const typeDecl = resolveQualifiedType(
      typeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!typeDecl) {
      throw new SyntaxError(
        `type argument passed to ${fnName}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    if (type === 'props') propsTypeDecl = typeDecl
    else emitsTypeDecl = typeDecl

    if (declId) {
      if (type === 'props' && declId.type === 'ObjectPattern') {
        propsDestructureDecl = declId
      } else if (type === 'emits' && declId.type === 'Identifier') {
        emitsIdentifier = declId.name
      }
    } else if (type === 'emits') {
      emitsIdentifier = `_${DEFINE_MODEL}_emit`
      s.prependRight(setupOffset + node.start!, `const ${emitsIdentifier} = `)
    }

    return true
  }

  function processDefineModel(
    node: Node,
    declId?: LVal,
    kind?: VariableDeclaration['kind']
  ) {
    if (!isCallOf(node, DEFINE_MODEL)) {
      return false
    }

    if (hasDefineModel) {
      throw new SyntaxError(`duplicate ${DEFINE_MODEL}() call`)
    }
    hasDefineModel = true

    const propsTypeDeclRaw = node.typeParameters?.params[0]
    if (!propsTypeDeclRaw) {
      throw new SyntaxError(`expected a type parameter for ${DEFINE_MODEL}.`)
    }
    modelTypeDecl = resolveQualifiedType(
      propsTypeDeclRaw,
      (node) => node.type === 'TSTypeLiteral'
    ) as TSTypeLiteral | TSInterfaceBody | undefined

    if (!modelTypeDecl) {
      throw new SyntaxError(
        `type argument passed to ${DEFINE_MODEL}() must be a literal type, ` +
          `or a reference to an interface or literal type.`
      )
    }

    if (declId) {
      const ids = extractIdentifiers(declId)
      ids.forEach((id) => modelIdentifiers.add(id))

      if (declId.type === 'ObjectPattern') {
        modelDestructureDecl = declId
        for (const property of declId.properties) {
          if (property.type === 'RestElement') {
            throw new SyntaxError('rest element is not supported')
          }
        }
      } else {
        modelIdentifier = scriptCompiled.loc.source.slice(
          declId.start!,
          declId.end!
        )
      }
    }
    if (kind) modelDeclKind = kind

    return true
  }

  function processDefineOptions(node: Node) {
    if (!isCallOf(node, DEFINE_OPTIONS)) return false

    const [arg] = node.arguments
    if (arg) processVue2Model(arg)

    return true
  }

  function processVue2Model(node: Node) {
    // model: {
    //   prop: 'checked',
    //   event: 'change'
    // }
    if (node.type !== 'ObjectExpression') return false

    const model = node.properties.find(
      (prop) =>
        prop.type === 'ObjectProperty' &&
        prop.key.type === 'Identifier' &&
        prop.key.name === 'model' &&
        prop.value.type === 'ObjectExpression' &&
        prop.value.properties.length === 2
    ) as ObjectProperty

    if (!model) return false
    ;(model.value as ObjectExpression).properties.forEach((propertyItem) => {
      if (
        propertyItem.type === 'ObjectProperty' &&
        propertyItem.key.type === 'Identifier' &&
        propertyItem.value.type === 'StringLiteral' &&
        ['prop', 'event'].includes(propertyItem.key.name)
      ) {
        const key = propertyItem.key.name as 'prop' | 'event'
        modelVue2[key] = propertyItem.value.value
      }
    })
    return true
  }

  function resolveQualifiedType(
    node: Node,
    qualifier: (node: Node) => boolean
  ) {
    if (qualifier(node)) {
      return node
    }
    if (
      node.type === 'TSTypeReference' &&
      node.typeName.type === 'Identifier'
    ) {
      const refName = node.typeName.name
      const isQualifiedType = (node: Node): Node | undefined => {
        if (
          node.type === 'TSInterfaceDeclaration' &&
          node.id.name === refName
        ) {
          return node.body
        } else if (
          node.type === 'TSTypeAliasDeclaration' &&
          node.id.name === refName &&
          qualifier(node.typeAnnotation)
        ) {
          return node.typeAnnotation
        } else if (node.type === 'ExportNamedDeclaration' && node.declaration) {
          return isQualifiedType(node.declaration)
        }
      }
      const body = sfc.scriptCompiled.scriptSetupAst!
      for (const node of body) {
        const qualified = isQualifiedType(node)
        if (qualified) {
          return qualified
        }
      }
    }
  }

  function extractRuntimeProps(
    node: TSTypeLiteral | TSInterfaceBody
  ): Record<string, string> {
    const members = node.type === 'TSTypeLiteral' ? node.members : node.body
    const map: Record<string, string> = {}
    for (const m of members) {
      if (
        (m.type === 'TSPropertySignature' || m.type === 'TSMethodSignature') &&
        m.key.type === 'Identifier'
      ) {
        const value = scriptCompiled.loc.source.slice(
          m.typeAnnotation!.start!,
          m.typeAnnotation!.end!
        )
        map[m.key.name] = value
      }
    }
    return map
  }

  function getEventKey(key: string) {
    if (version === 2) {
      if (modelVue2.prop === key) {
        return modelVue2.event
      } else if (key === 'value') {
        return 'input'
      }
    }
    return `update:${key}`
  }

  function processAssignModelVariable() {
    if (!emitsIdentifier)
      throw new Error(
        `Identifier of returning value of ${DEFINE_EMITS} is not found, please report this issue.\n${REPO_ISSUE_URL}`
      )

    const program: Program = {
      type: 'Program',
      body: scriptCompiled.scriptSetupAst as Statement[],
      directives: [],
      sourceType: 'module',
      sourceFile: '',
    }
    let hasTransfromed = false

    function overwrite(
      node: Node,
      id: Identifier,
      value: string,
      original = false
    ) {
      hasTransfromed = true
      const content = `__emitHelper(${emitsIdentifier}, '${getEventKey(
        id.name
      )}', ${value}${original ? `, ${id.name}` : ''})`
      s.overwrite(setupOffset + node.start!, setupOffset + node.end!, content)
    }

    walkAST(program, {
      leave(node) {
        if (node.type === 'AssignmentExpression') {
          if (node.left.type !== 'Identifier') return
          const id = this.scope[node.left.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          const left = s.sliceNode(node.left, { offset: setupOffset })
          let right = s.sliceNode(node.right, { offset: setupOffset })
          if (node.operator !== '=') {
            right = `${left} ${node.operator.replace(/=$/, '')} ${right}`
          }

          overwrite(node, id, right)
        } else if (node.type === 'UpdateExpression') {
          if (node.argument.type !== 'Identifier') return
          const id = this.scope[node.argument.name] as Identifier
          if (!modelIdentifiers.has(id)) return

          let value = node.argument.name
          if (node.operator === '++') value += ' + 1'
          else value += ' - 1'

          overwrite(node, id, value, !node.prefix)
        }
      },
    })

    if (hasTransfromed) {
      s.prependLeft(
        setupOffset,
        `\nimport __emitHelper from '${emitHelperId}';`
      )
    }
  }

  if (!code.includes(DEFINE_MODEL)) return
  const sfc = parseSFC(code, id)
  if (!sfc.scriptSetup) return

  const { scriptCompiled } = sfc
  if (!scriptCompiled) return

  const s = new MagicString(code)
  const setupOffset = scriptCompiled.loc.start.offset

  if (
    version === 2 &&
    scriptCompiled.scriptAst &&
    scriptCompiled.scriptAst.length > 0
  ) {
    // process normal <script>
    for (const node of scriptCompiled.scriptAst as Statement[]) {
      if (node.type === 'ExportDefaultDeclaration') {
        const { declaration } = node
        if (declaration.type === 'ObjectExpression') {
          processVue2Model(declaration)
        } else if (
          declaration.type === 'CallExpression' &&
          declaration.callee.type === 'Identifier' &&
          declaration.callee.name === 'defineComponent'
        ) {
          declaration.arguments.forEach((arg) => {
            if (arg.type === 'ObjectExpression') {
              processVue2Model(arg)
            }
          })
        }
      }
    }
  }

  // process <script setup>
  for (const node of scriptCompiled.scriptSetupAst as Statement[]) {
    if (node.type === 'ExpressionStatement') {
      processDefinePropsOrEmits(node.expression)

      if (version === 2) {
        processDefineOptions(node.expression)
      }

      if (processDefineModel(node.expression)) {
        s.remove(node.start! + setupOffset, node.end! + setupOffset)
      }
    } else if (node.type === 'VariableDeclaration' && !node.declare) {
      const total = node.declarations.length
      let left = total

      for (let i = 0; i < total; i++) {
        const decl = node.declarations[i]
        if (decl.init) {
          processDefinePropsOrEmits(decl.init, decl.id)

          if (processDefineModel(decl.init, decl.id, node.kind)) {
            if (left === 1) {
              s.remove(node.start! + setupOffset, node.end! + setupOffset)
            } else {
              let start = decl.start! + setupOffset
              let end = decl.end! + setupOffset
              if (i < total - 1) {
                // not the last one, locate the start of the next
                end = node.declarations[i + 1].start! + setupOffset
              } else {
                // last one, locate the end of the prev
                start = node.declarations[i - 1].end! + setupOffset
              }
              s.remove(start, end)
              left--
            }
          }
        }
      }
    }
  }

  if (!modelTypeDecl) return
  if (modelTypeDecl.type !== 'TSTypeLiteral') {
    throw new SyntaxError(
      `type argument passed to ${DEFINE_MODEL}() must be a literal type, or a reference to an interface or literal type.`
    )
  }

  const map = extractRuntimeProps(modelTypeDecl)

  const propsText = Object.entries(map)
    .map(([key, value]) => `${key}${value}`)
    .join('\n')

  const emitsText = Object.entries(map)
    .map(([key, value]) => `(evt: '${getEventKey(key)}', value${value}): void`)
    .join('\n')

  if (hasDefineProps) {
    s.appendLeft(setupOffset + propsTypeDecl!.start! + 1, `${propsText}\n`)
    if (propsDestructureDecl && modelDestructureDecl)
      for (const property of modelDestructureDecl.properties) {
        const text = code.slice(
          setupOffset + property.start!,
          setupOffset + property.end!
        )
        s.appendLeft(setupOffset + propsDestructureDecl.start! + 1, `${text}, `)
      }
  } else {
    let text = ''
    const kind = modelDeclKind || 'let'
    if (modelIdentifier) {
      text = modelIdentifier
    } else if (modelDestructureDecl) {
      text = code.slice(
        setupOffset + modelDestructureDecl.start!,
        setupOffset + modelDestructureDecl.end!
      )
    }

    s.appendLeft(
      setupOffset,
      `\n${text ? `${kind} ${text} = ` : ''}defineProps<{
  ${propsText}
}>();`
    )
  }
  if (hasDefineEmits) {
    s.appendLeft(setupOffset + emitsTypeDecl!.start! + 1, `${emitsText}\n`)
  } else {
    emitsIdentifier = `_${DEFINE_MODEL}_emit`
    s.appendLeft(
      setupOffset,
      `\nconst ${emitsIdentifier} = defineEmits<{
  ${emitsText}
}>();`
    )
  }

  if (hasDefineModel) {
    processAssignModelVariable()
  }

  return getTransformResult(s, id)
}
