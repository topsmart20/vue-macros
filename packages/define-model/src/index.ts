import { createUnplugin } from 'unplugin'
import {
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
  normalizePath,
} from '@vue-macros/common'
import { transformDefineModel } from './core'
import {
  emitHelperCode,
  emitHelperId,
  helperPrefix,
  useVmodelHelperCode,
  useVmodelHelperId,
} from './core/helper'
import type { BaseOptions, MarkRequired } from '@vue-macros/common'
import type { UnpluginContextMeta } from 'unplugin'

export interface Options extends BaseOptions {
  /**
   * Unified mode, only works for Vue 2
   *
   * Converts `modelValue` to `value`
   */
  unified?: boolean
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'unified'
>

function resolveOption(
  options: Options,
  framework: UnpluginContextMeta['framework']
): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [REGEX_VUE_SFC, REGEX_SETUP_SFC].concat(
      version === 2 && framework === 'webpack' ? REGEX_VUE_SUB : []
    ),
    unified: true,
    ...options,
    version,
  }
}

const name = 'unplugin-vue-define-model'

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOption(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id.startsWith(helperPrefix)) return id
      },

      loadInclude(id) {
        return id.startsWith(helperPrefix)
      },

      load(_id) {
        const id = normalizePath(_id)
        if (id === emitHelperId) return emitHelperCode
        else if (id === useVmodelHelperId) return useVmodelHelperCode
      },

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineModel(code, id, options.version, options.unified)
      },
    }
  }
)
