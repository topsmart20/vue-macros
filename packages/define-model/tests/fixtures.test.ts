import { describe, expect, test } from 'vitest'
import { transformDefineModel } from '../src/core'

describe('fixtures', async () => {
  const files = import.meta.glob('./fixtures/**/*.{vue,js,ts}', {
    eager: true,
    as: 'raw',
  })

  for (const [id, code] of Object.entries(files)) {
    test(id.replace(/\\/g, '/'), async () => {
      const version = id.includes('vue2') ? 2 : 3
      const exec = () => transformDefineModel(code, id, version)?.code
      if (id.includes('error')) {
        expect(exec).toThrowErrorMatchingSnapshot()
      } else {
        expect(exec()).toMatchSnapshot()
      }
    })
  }
})
