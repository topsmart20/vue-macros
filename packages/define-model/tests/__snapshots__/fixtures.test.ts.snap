// Vitest Snapshot v1

exports[`fixtures > ./fixtures/reactivity-transform/assignment.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_emitHelper from '/plugin-define-model/emit-helper';
let { modelValue, title } = defineProps<{
    modelValue: string
title: string
  }>();
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:title', value: string): void
  }>();


_DM_emitHelper(_DM_emit, 'update:modelValue', _DM_emitHelper(_DM_emit, 'update:title', '123'))
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/assignment-update.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_emitHelper from '/plugin-define-model/emit-helper';
let { count } = defineProps<{
    count: number
  }>();
const _DM_emit = defineEmits<{
    (evt: 'update:count', value: number): void
  }>();


_DM_emitHelper(_DM_emit, 'update:count', count + 1, count)
_DM_emitHelper(_DM_emit, 'update:count', count - 1, count)
_DM_emitHelper(_DM_emit, 'update:count', count + 1)
_DM_emitHelper(_DM_emit, 'update:count', count - 1)
_DM_emitHelper(_DM_emit, 'update:count', 1)
_DM_emitHelper(_DM_emit, 'update:count', count + 1)
_DM_emitHelper(_DM_emit, 'update:count', count - 1)
_DM_emitHelper(_DM_emit, 'update:count', count * 1)
_DM_emitHelper(_DM_emit, 'update:count', count / 1)

_DM_emitHelper(_DM_emit, 'update:count', count + 1)

const additionAssignment = () => (_DM_emitHelper(_DM_emit, 'update:count', count + 1))
const increment = () => _DM_emitHelper(_DM_emit, 'update:count', count + 1, count)
const add = () => (_DM_emitHelper(_DM_emit, 'update:count', count + 1))
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_emitHelper from '/plugin-define-model/emit-helper';
var { modelValue } = defineProps<{
    modelValue: string
visible: boolean
  }>();
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();


console.log(modelValue)
_DM_emitHelper(_DM_emit, 'update:modelValue', 'newValue')
</script>

<template>
  <h1>123</h1>
</template>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/error-rest.vue 1`] = `"rest element is not supported"`;

exports[`fixtures > ./fixtures/reactivity-transform/multi-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
let { modelValue } = defineProps<{
    modelValue: string
visible: boolean
  }>();
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
let a = 'b'
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/no-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
defineProps<{
    modelValue: string
visible: boolean
  }>();
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();

</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-emits.vue 1`] = `
"<script setup lang=\\"ts\\">
const { modelValue } = defineProps<{
    modelValue: string
  }>();


const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

emit('change')
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-emits-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_emitHelper from '/plugin-define-model/emit-helper';
let { modelValue } = defineProps<{
    modelValue: string
  }>();


const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

{
  _DM_emitHelper(emit, 'update:modelValue', 'hello')
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-emits-no-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_emitHelper from '/plugin-define-model/emit-helper';
let { modelValue } = defineProps<{
    modelValue: string
  }>();


const _defineModel_emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

_DM_emitHelper(_defineModel_emit, 'update:modelValue', '1')
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-props.vue 1`] = `
"<script setup lang=\\"ts\\">
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();


defineProps<{modelValue: string
visible: boolean

  title?: string
}>()
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-props-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();


const {modelValue,  title } = defineProps<{modelValue: string
visible: boolean

  title?: string
}>()
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-props-empty.vue 1`] = `
"<script setup lang=\\"ts\\">
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();


defineProps< {modelValue: string
visible: boolean
  } > ()
</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-props-ref.vue 1`] = `
"<script setup lang=\\"ts\\">
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
interface Props {modelValue: string
visible: boolean

  title?: string
}

defineProps<Props>()


</script>
"
`;

exports[`fixtures > ./fixtures/reactivity-transform/with-define-props-rest.vue 1`] = `
"<script setup lang=\\"ts\\">
const _DM_emit = defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();


const {modelValue,  title, ...props } = defineProps<{modelValue: string
visible: boolean

  title?: string
}>()

console.log(props)
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue2/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
let { value, title } = _DM_useVModel(['value', 'value', 'input'], ['title', 'title', 'update:title'])

const { id } = defineProps<{value: string
title: string
 id: string }>()

const emit = defineEmits<{(evt: 'input', value: string): void
(evt: 'update:title', value: string): void

  (evt: 'change'): void
}>()

{
  value = 'hello'
  title = 'world'
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue2/custom-model.vue 1`] = `
"<script lang=\\"ts\\">
export default {
  model: {
    prop: 'value',
    event: 'changeInput',
  },
}
</script>
<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
let { value, title } = _DM_useVModel(['value', 'value', 'changeInput'], ['title', 'title', 'update:title'])

const { id } = defineProps<{value: string
title: string
 id: string }>()

const emit = defineEmits<{(evt: 'changeInput', value: string): void
(evt: 'update:title', value: string): void

  (evt: 'change'): void
}>()

{
  value = 'hello'
  title = 'world'
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue2/custom-model-defineComponent.vue 1`] = `
"<script lang=\\"ts\\">
import { defineComponent } from 'vue'
export default defineComponent({
  model: {
    prop: 'value',
    event: 'changeInput'
  },
})
</script>
<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
let { value, title } = _DM_useVModel(['value', 'value', 'changeInput'], ['title', 'title', 'update:title'])

const { id } = defineProps<{value: string
title: string
 id: string }>()

const emit = defineEmits<{(evt: 'changeInput', value: string): void
(evt: 'update:title', value: string): void

  (evt: 'change'): void
}>()

{
  value = 'hello'
  title = 'world'
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue2/v-model.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    value: string
  }>();
defineEmits<{
    (evt: 'input', value: string): void
  }>();
const { modelValue } = _DM_useVModel(['modelValue', 'value', 'input'])

modelValue.value = 'hello'
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/assignment.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
title: string
  }>();
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:title', value: string): void
  }>();
const { modelValue, title } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['title', 'title', 'update:title'])

modelValue.value = title.value = '123'
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/basic.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
visible: boolean
  }>();
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
const { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])

console.log(modelValue.value)
modelValue.value = 'newValue'
</script>

<template>
  <h1>123</h1>
</template>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/define-props-runtime.vue 1`] = `undefined`;

exports[`fixtures > ./fixtures/runtime/vue3/empty.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    
  }>();
defineEmits<{
    
  }>();
_DM_useVModel()
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/error-empty-define-props.vue 1`] = `"defineProps() expected a type parameter when used with defineModel."`;

exports[`fixtures > ./fixtures/runtime/vue3/error-runtime.vue 1`] = `"defineProps() cannot accept non-type arguments when used with defineModel()"`;

exports[`fixtures > ./fixtures/runtime/vue3/multi-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
visible: boolean
  }>();
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
const a = 'b',
  { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/no-decl.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
visible: boolean
  }>();
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
_DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/rest.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
visible: boolean
  }>();
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
const { ...rest } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])

rest.modelValue.value = 'foo'
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/with-defaults.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineEmits<{
    (evt: 'update:foo', value: string): void
  }>();
export interface Props {foo: string

  color: string
}

const { foo } = _DM_useVModel(['foo', 'foo', 'update:foo'])

withDefaults(defineProps<Props>(), {
  color: 'blue'
})

console.log(foo)
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/with-define-emits.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineProps<{
    modelValue: string
  }>();
const { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'])

const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

emit('change')
modelValue.value = 'hello'
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/with-define-props.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
const { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])

defineProps<{modelValue: string
visible: boolean

  title?: string
}>()
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/with-define-props-empty.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
const { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])

defineProps< {modelValue: string
visible: boolean
  } > ()
</script>
"
`;

exports[`fixtures > ./fixtures/runtime/vue3/with-define-props-ref.vue 1`] = `
"<script setup lang=\\"ts\\">
import _DM_useVModel from '/plugin-define-model/use-vmodel';
defineEmits<{
    (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
  }>();
interface Props {modelValue: string
visible: boolean

  title?: string
}

defineProps<Props>()

const { modelValue } = _DM_useVModel(['modelValue', 'modelValue', 'update:modelValue'], ['visible', 'visible', 'update:visible'])
</script>
"
`;
